blog/2021-06-29-transaction-presigning.md

```diff
@@ -0,0 +1,148 @@
+Unfortunately, it only supports selling BTC for XMR from the user's perspective.
+
+To explain why, we need to dive a bit further into the workings of atomic swap protocols.
+From a high-level perspective, an atomic swap consists of four key actions:
+
+1. Party A locks up funds
+1. Party B locks up funds
+1. Party A takes party B's funds
+1. Party B takes party A's funds
+
+To build a non-custodial XMR-BTC exchange service, you really want to be party B.
+Service meaning, you are running some form of automated backend that allows users to initiate and complete a swap without direct engagement from your side.
+
+If you were offering such a service as party A, people could come and initiate a swap (which means you have to lock up money) but then just walk away. 
+Even if you can take your funds back due to a timeout or some other mechanism, your service is subject to a very easy DoS where the fees end up eating up your balance.
+
+A protocol where Party A locks up BTC, followed by party B locking up XMR is what is currently implemented [in our software](https://github.com/comit-network/xmr-btc-swap).
+Here, the user is limited to buying XMR for BTC (or selling BTC for XMR, depending on how you look at it).
+
+If we want to allow users to sell XMR for BTC, then we need a protocol where party A locks up XMR first. 
+We [developed such a protocol](https://arxiv.org/abs/2101.12332) but in order for it to be secure from a game-theoretic PoV, we need to pre-sign spending transaction from a Monero joint output.
+
+In a nutshell, if we want to provide a non-custodial trading solution where users can buy AND sell XMR for BTC, we need transaction pre-signing.
+
+## It doesn't work on present day Monero
+
+Unfortunately, it is not possible to efficiently implement pre-signing of transactions in present day Monero. 
+The reason for that is very technical and we believe it can be fixed but likely requires a hard-fork.
+
+Lucas recently [made an effort](/blog/2021/05/19/monero-transaction) in documenting the transaction structure and signature algorithms of Monero.
+I'll try to reiterate the necessary concepts here but if you want more information on how things work, I suggest to take a look at Lucas' post.
+
+To sign a transaction in Monero, one needs to choose ten other outputs that - together with the output you want to spend - form the ring for Monero's CLSAG algorithm.
+These outputs are represented as what is called *key-offsets* for efficiency reasons.
+For the ring itself, we need the public key of each output but public keys are relatively big in size.
+32 bytes even in their compressed form to be precise.
+Times 11, that would be 352 bytes *per input* just to represent the ring for the signature.
+
+To be more space efficient, we instead enumerate all outputs in the blockchain and create a list of offsets where the first element in the list is an absolute output index and all subsequent elements are relative to their previous index.
```

Comment:
nit: `To be more space efficient, we instead enumerate` -> I find this "we" a bit misleading. This refers to the present day Monero implementation if I understand it correctly and this we is a general "we" as in "one". Maybe just change to "Monero"?

Reply:
I don't know how to explain this but I believe it is a fairly common style in describing algorithms and/or approaches to say `we`.

It is a nice way of including the reader in the walk-through of the solution.

Maybe @luckysori can explain this better.